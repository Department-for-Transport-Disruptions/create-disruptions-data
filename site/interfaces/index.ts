import { NextApiRequest } from "next";
import { z } from "zod";
import { ServerResponse } from "http";
import { Service, Stop } from "../schemas/consequence.schema";
import { ConsequenceType } from "../schemas/type-of-consequence.schema";

export interface ErrorInfo {
    errorMessage: string;
    id: string;
    userInput?: string;
}

export interface ResponseWithLocals extends ServerResponse {
    locals: {
        nonce: string;
        csrfToken: string;
    };
}

export interface CookiePolicy {
    essential: boolean;
    usage: boolean;
}

export interface CookiesApiRequest extends NextApiRequest {
    body: {
        tracking: "on" | "off";
    };
}

export interface FormBase<T> {
    value?: string;
    display: string;
    displaySize?: "s" | "m" | "l" | "xl";
    inputName: Extract<keyof T, string>;
    initialErrors?: ErrorInfo[];
    stateUpdater: (change: string, field: keyof T) => void;
    schema?: z.ZodTypeAny;
}

export interface DisplayValuePair<T = string> {
    display: string;
    value: T;
    checked?: boolean;
}

export interface Feedback {
    question: string;
    answer: string;
}

export interface TestInputs {
    field1: string;
    field2: boolean;
    field3: string;
    field4: string;
}

export interface PageState<T> {
    errors: ErrorInfo[];
    inputs: T;
}

export interface SocialMediaPost {
    messageToAppear: string;
    publishDate: string;
    publishTime: string;
    accountToPublish: string;
}

export interface CreateConsequenceProps<T> {
    initialPageState: PageState<Partial<T>>;
    previousConsequenceInformation: ConsequenceType;
    initialServices?: Service[];
    initialStops?: Stop[];
    allStops?: Stop[]
}
