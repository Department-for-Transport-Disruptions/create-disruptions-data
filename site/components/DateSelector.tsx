import React, { ReactElement } from 'react';
import { ErrorInfo } from '../interfaces';
import FormElementWrapper from './FormElementWrapper';
import camelCase from 'lodash/camelCase';
interface DateSelectorProps {
    inputs?: {
        dayInput: string;
        monthInput: string;
        yearInput: string;
    };
    errors?: ErrorInfo[];
    startOrEnd: 'start' | 'end';
    type?: string
}

const DateSelector = ({ inputs, startOrEnd, errors = [], type }: DateSelectorProps): ReactElement => (
    <FormElementWrapper errors={errors} errorId={`${startOrEnd}-day-input`} errorClass="govuk-date-input--error">
        <div className="govuk-date-input" id={`${startOrEnd}-date`}>
            <div className="govuk-date-input__item">
                <div className="govuk-form-group">
                    <label className="govuk-label govuk-date-input__label" htmlFor={`${startOrEnd}-day-input`}>
                        Day
                    </label>
                    <input
                        className={`govuk-input govuk-date-input__input govuk-input--width-2 ${errors.length > 0 ? 'govuk-input--error' : ''
                            } `}
                        id={`${startOrEnd}-day-input`}
                        name={`${startOrEnd}DateDay`}
                        type="text"
                        defaultValue={inputs?.dayInput}
                    />
                </div>
            </div>

            <div className="govuk-date-input__item">
                <div className="govuk-form-group">
                    <label className="govuk-label govuk-date-input__label" htmlFor={`${startOrEnd}-month-input`}>
                        Month
                    </label>
                    <input
                        className={`govuk-input govuk-date-input__input govuk-input--width-2 ${errors.length > 0 ? 'govuk-input--error' : ''
                            } `}
                        id={`${type}-${startOrEnd}-month-input`}
                        name={`${type}${type ? camelCase(startOrEnd) : startOrEnd}DateMonth`}
                        type="text"
                        defaultValue={inputs?.monthInput}
                    />
                </div>
            </div>
            <div className="govuk-date-input__item">
                <div className="govuk-form-group">
                    <label className="govuk-label govuk-date-input__label" htmlFor={`${startOrEnd}-year-input`}>
                        Year
                    </label>
                    <input
                        className={`govuk-input govuk-date-input__input govuk-input--width-4 ${errors.length > 0 ? 'govuk-input--error' : ''
                            } `}
                        id={`$${type}-${startOrEnd}-year-input`}
                        name={`${type}${type ? camelCase(startOrEnd) : startOrEnd}DateYear`}
                        type="text"
                        defaultValue={inputs?.yearInput}
                    />
                </div>
            </div>
        </div>
    </FormElementWrapper>
);

export default DateSelector;
