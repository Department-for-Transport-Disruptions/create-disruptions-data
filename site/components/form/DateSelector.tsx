import { FilledInputProps } from "@mui/material/FilledInput";
import { InputBaseComponentProps } from "@mui/material/InputBase";
import { OutlinedInputProps } from "@mui/material/OutlinedInput";
import { DatePicker, PickersDay, PickersDayProps } from "@mui/x-date-pickers";
import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
import { LocalizationProvider } from "@mui/x-date-pickers/LocalizationProvider";
import kebabCase from "lodash/kebabCase";
import React, { ReactElement, useEffect, useState } from "react";
import { z } from "zod";
import FormElementWrapper, { FormGroupWrapper } from "./FormElementWrapper";
import { ErrorInfo, FormBase } from "../../interfaces";
import { convertDateTimeToFormat, getFormattedDate } from "../../utils/dates";
import { handleBlur } from "../../utils/formUtils";

interface DateSelectorProps<T> extends FormBase<T> {
    disabled: boolean;
    hiddenHint?: string;
    disablePast: boolean;
    reset?: boolean;
}

const inputBox = <T extends object>(
    inputRef: React.Ref<HTMLInputElement> | undefined,
    inputProps: InputBaseComponentProps | undefined,
    InputProps: Partial<FilledInputProps> | Partial<OutlinedInputProps> | undefined,
    inputId: string,
    inputName: Extract<keyof T, string>,
    errors: ErrorInfo[],
    disabled: boolean,
    stateUpdater: (change: string, field: keyof T) => void,
    setErrors: React.Dispatch<React.SetStateAction<ErrorInfo[]>>,
    schema?: z.ZodTypeAny,
) => (
    <div className="govuk-date-input flex items-center [&_.MuiSvgIcon-root]:fill-govBlue">
        <div className="govuk-date-input__item govuk-!-margin-right-0">
            <FormElementWrapper errors={errors} errorId={inputName} errorClass="govuk-input--error">
                <input
                    className="govuk-input govuk-date-input__input govuk-input--width-6"
                    name={inputName}
                    id={`${inputId}-input`}
                    type="text"
                    ref={inputRef}
                    {...inputProps}
                    disabled={disabled}
                    placeholder={disabled ? "N/A" : "DD/MM/YYYY"}
                    onBlur={(e) => handleBlur(e.target.value, inputName, stateUpdater, setErrors, schema, disabled)}
                />
            </FormElementWrapper>
        </div>
        {InputProps?.endAdornment}
    </div>
);

const renderWeekPickerDay = (
    _date: Date,
    _selectedDates: Array<Date | null>,
    pickersDayProps: PickersDayProps<Date>,
) => (
    <PickersDay
        {...pickersDayProps}
        classes={{
            selected: "!bg-govBlue",
            dayWithMargin:
                "focus:!border focus:!border-solid hover:!border hover:!border-solid hover:!border-govBlue focus:!border-govYellow",
        }}
    />
);

const DateSelector = <T extends object>({
    value,
    display,
    displaySize = "s",
    inputName,
    initialErrors = [],
    disabled,
    hiddenHint,
    disablePast,
    stateUpdater,
    schema,
    reset = false,
}: DateSelectorProps<T>): ReactElement => {
    const [dateValue, setDateValue] = useState<Date | null>(
        !!disabled || !value ? null : getFormattedDate(value).toDate(),
    );
    const [errors, setErrors] = useState<ErrorInfo[]>(initialErrors);
    const inputId = kebabCase(inputName);

    useEffect(() => {
        if (disabled || reset) {
            setErrors([]);
            setDateValue(null);
        }
    }, [disabled, reset]);

    useEffect(() => {
        setErrors(initialErrors);
    }, [initialErrors]);

    return (
        <FormGroupWrapper errorIds={[inputName]} errors={errors}>
            <div className="govuk-form-group govuk-!-margin-bottom-0" id={inputId}>
                <label className={`govuk-label govuk-label--${displaySize}`} htmlFor={`${inputId}-input`}>
                    {display}
                </label>
                {hiddenHint ? <div className="govuk-hint govuk-visually-hidden">{hiddenHint}</div> : null}
                <LocalizationProvider dateAdapter={AdapterDayjs}>
                    <DatePicker
                        renderDay={renderWeekPickerDay}
                        value={dateValue}
                        onChange={(newValue) => {
                            setDateValue(newValue);
                            if (newValue) {
                                stateUpdater(convertDateTimeToFormat(newValue, "DD/MM/YYYY"), inputName);
                            }
                        }}
                        renderInput={({ inputRef, inputProps, InputProps }) => {
                            return inputBox(
                                inputRef,
                                inputProps,
                                InputProps,
                                inputId,
                                inputName,
                                errors,
                                disabled,
                                stateUpdater,
                                setErrors,
                                schema,
                            );
                        }}
                        disablePast={disablePast}
                        inputFormat="DD/MM/YYYY"
                        disabled={disabled}
                    />
                </LocalizationProvider>
            </div>
        </FormGroupWrapper>
    );
};

export default DateSelector;
