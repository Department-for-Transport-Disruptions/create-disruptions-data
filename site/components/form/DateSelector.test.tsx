import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import renderer from "react-test-renderer";
import { describe, expect, it, vi } from "vitest";
import { TestInputs } from "../../interfaces";
import DateSelector from "./DateSelector";

describe("DateSelector", () => {
    it("should render correctly with no input", () => {
        const tree = renderer
            .create(
                <DateSelector<TestInputs>
                    display="Start date"
                    hint={{ hidden: false, text: "Enter in format DD/MM/YYYY" }}
                    value=""
                    disabled={false}
                    disablePast={false}
                    inputName="field1"
                    stateUpdater={vi.fn()}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly with no input and the hint is hidden", () => {
        const tree = renderer
            .create(
                <DateSelector<TestInputs>
                    display="Start date"
                    hint={{ hidden: true, text: "Enter in format DD/MM/YYYY" }}
                    value=""
                    disabled={false}
                    disablePast={false}
                    inputName="field1"
                    stateUpdater={vi.fn()}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly with errors", () => {
        const tree = renderer
            .create(
                <DateSelector<TestInputs>
                    display="Start date"
                    hint={{ hidden: false, text: "Enter in format DD/MM/YYYY" }}
                    initialErrors={[{ errorMessage: "There was an error", id: "field1" }]}
                    value="sss"
                    disabled={false}
                    disablePast={false}
                    inputName="field1"
                    stateUpdater={vi.fn()}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly with an input", () => {
        const tree = renderer
            .create(
                <DateSelector<TestInputs>
                    display="Start date"
                    hint={{ hidden: false, text: "Enter in format DD/MM/YYYY" }}
                    value="01/01/2024"
                    disabled={false}
                    disablePast={false}
                    inputName="field1"
                    stateUpdater={vi.fn()}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly when disabled", () => {
        const tree = renderer
            .create(
                <DateSelector<TestInputs>
                    display="Start date"
                    hint={{ hidden: false, text: "Enter in format DD/MM/YYYY" }}
                    value="01/01/2024"
                    disabled
                    disablePast={false}
                    inputName="field1"
                    stateUpdater={vi.fn()}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should not error when the component is disabled", async () => {
        const { unmount } = render(
            <DateSelector<TestInputs>
                display="Start date"
                hint={{ hidden: false, text: "Enter in format DD/MM/YYYY" }}
                value=""
                disabled
                disablePast={false}
                inputName="field1"
                stateUpdater={vi.fn()}
            />,
        );

        await userEvent.click(screen.getByLabelText("Start date"));
        await userEvent.tab();

        expect(screen.queryByText("Select a date")).toBeFalsy();

        unmount();
    });
});
