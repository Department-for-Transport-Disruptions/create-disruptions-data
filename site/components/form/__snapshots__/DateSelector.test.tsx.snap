// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DateSelector > should render correctly when disabled 1`] = `
<div
  className="govuk-form-group"
>
  <div
    className="govuk-form-group govuk-!-margin-bottom-0"
    id="field-1"
  >
    <label
      className="govuk-label govuk-label--s"
      htmlFor="field-1-input"
    >
      Start date
    </label>
    <div
      className="govuk-date-input flex items-center [&_.MuiSvgIcon-root]:fill-govBlue"
    >
      <div
        className="govuk-date-input__item govuk-!-margin-right-0"
      >
        <div
          className=""
        >
          <input
            className="govuk-input govuk-date-input__input govuk-input--width-6"
            disabled={true}
            id="field-1-input"
            name="field1"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="N/A"
            type="tel"
            value=""
          />
        </div>
      </div>
      <div
        className="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
      >
        <button
          aria-label="Choose date"
          className="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CalendarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`DateSelector > should render correctly with an input 1`] = `
<div
  className="govuk-form-group"
>
  <div
    className="govuk-form-group govuk-!-margin-bottom-0"
    id="field-1"
  >
    <label
      className="govuk-label govuk-label--s"
      htmlFor="field-1-input"
    >
      Start date
    </label>
    <div
      className="govuk-date-input flex items-center [&_.MuiSvgIcon-root]:fill-govBlue"
    >
      <div
        className="govuk-date-input__item govuk-!-margin-right-0"
      >
        <div
          className=""
        >
          <input
            className="govuk-input govuk-date-input__input govuk-input--width-6"
            disabled={false}
            id="field-1-input"
            name="field1"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="DD/MM/YYYY"
            type="tel"
            value="01/01/2024"
          />
        </div>
      </div>
      <div
        className="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
      >
        <button
          aria-label="Choose date, selected date is Jan 1, 2024"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CalendarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`DateSelector > should render correctly with errors 1`] = `
<div
  className="govuk-form-group govuk-form-group--error"
>
  <div
    className="govuk-form-group govuk-!-margin-bottom-0"
    id="field-1"
  >
    <label
      className="govuk-label govuk-label--s"
      htmlFor="field-1-input"
    >
      Start date
    </label>
    <div
      className="govuk-date-input flex items-center [&_.MuiSvgIcon-root]:fill-govBlue"
    >
      <div
        className="govuk-date-input__item govuk-!-margin-right-0"
      >
        <div
          className=""
        >
          <span
            className="govuk-error-message"
            id="field1-error"
          >
            <span
              className="govuk-visually-hidden"
            >
              Error: 
            </span>
            There was an error
          </span>
          <input
            aria-describedby="field1-error"
            className="govuk-input govuk-date-input__input govuk-input--width-6 govuk-input--error"
            disabled={false}
            id="field-1-input"
            name="field1"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="DD/MM/YYYY"
            type="tel"
            value=""
          />
        </div>
      </div>
      <div
        className="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
      >
        <button
          aria-label="Choose date"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CalendarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
            />
          </svg>
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`DateSelector > should render correctly with no input 1`] = `
<div
  className="govuk-form-group"
>
  <div
    className="govuk-form-group govuk-!-margin-bottom-0"
    id="field-1"
  >
    <label
      className="govuk-label govuk-label--s"
      htmlFor="field-1-input"
    >
      Start date
    </label>
    <div
      className="govuk-date-input flex items-center [&_.MuiSvgIcon-root]:fill-govBlue"
    >
      <div
        className="govuk-date-input__item govuk-!-margin-right-0"
      >
        <div
          className=""
        >
          <input
            className="govuk-input govuk-date-input__input govuk-input--width-6"
            disabled={false}
            id="field-1-input"
            name="field1"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="DD/MM/YYYY"
            type="tel"
            value=""
          />
        </div>
      </div>
      <div
        className="MuiInputAdornment-root MuiInputAdornment-positionEnd css-1laqsz7-MuiInputAdornment-root"
      >
        <button
          aria-label="Choose date"
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="CalendarIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
